File: 백준/Silver/2583. 영역 구하기/영역 구하기.py
import sys
sys.setrecursionlimit(10**6)
input = sys.stdin.readline

dx = [-1, 1, 0, 0]
dy = [0, 0, -1, 1]


def dfs(graph, x, y):
    count = 1
    graph[x][y] = 1

    for i in range(4):
        nx = x + dx[i]
        ny = y + dy[i]
        if 0 <= nx < len(graph) and 0 <= ny < len(graph[0]) and graph[nx][ny] == 0:
            count += dfs(graph, nx, ny)

    return count


# m행 n열
m, n, k = map(int, input().split())

graph = [[0] * n for _ in range(m)]

for _ in range(k):
    # 왼 하단, 우 상단
    x1, y1, x2, y2 = map(int, input().split())
    for i in range(y1, y2):
        for j in range(x1, x2):
            graph[i][j] = 1

result_space = []

for i in range(m):
    for j in range(n):
        if graph[i][j] == 0:
            # 빈칸이면 개수 체크
            result_space.append(dfs(graph, i, j))

print(len(result_space))
result_space.sort()
print(' '.join(map(str, result_space)))

File: 백준/Silver/2583. 영역 구하기/영역 구하기.kt
package org.example

import java.util.StringTokenizer

private val dx = intArrayOf(-1, 1, 0, 0)
private val dy = intArrayOf(0, 0, -1, 1)

fun main() {
    val br = System.`in`.bufferedReader()
    val (N, M, K) = br.readLine().split(" ").map { it.toInt() }
    val graph = Array(M) { IntArray(N) }
    repeat(K) {
        val st = StringTokenizer(br.readLine())
        val x1 = st.nextToken().toInt()
        val y1 = st.nextToken().toInt()
        val x2 = st.nextToken().toInt()
        val y2 = st.nextToken().toInt()

        for (i in x1 until x2) {
            for (j in y1 until y2) {
                //println("$i, $j")
                graph[i][j] = 1 // 네모 영역 표시
            }
        }
    }

    val areaCount = mutableListOf<Int>()
    for (i in graph.indices) {
        for (j in graph[0].indices) {
            if (graph[i][j] == 0) {
                areaCount.add(dfs(graph, i, j))
            }
        }
    }

    areaCount.sort()
    val sb = StringBuilder()
    sb.append(areaCount.size).append("\n")
    areaCount.forEach { sb.append(it).append(" ") }
    sb.deleteCharAt(sb.length - 1)
    println(sb)

    br.close()
}

fun dfs(graph: Array<IntArray>, x: Int, y: Int): Int {
    var count = 1
    graph[x][y] = 2
    for (i in 0 until 4) {
        val nx = x + dx[i]
        val ny = y + dy[i]
        if (nx in graph.indices && ny in graph[0].indices && graph[nx][ny] == 0) {
            count += dfs(graph, nx, ny)
        }
    }
    return count
}
